include/ofi_coll.h

```diff
@@ -104,52 +109,67 @@ struct util_coll_reduce_item {
 	enum fi_op		op;
 };
 
-struct util_coll_comp_item;
-
-typedef void (*util_coll_comp_t)(struct util_coll_mc *coll_mc,
-				 struct util_coll_comp_item *comp);
-
-struct util_coll_join_comp_data {
-	uint64_t		cid_buf[OFI_CONTEXT_ID_SIZE];
-	uint64_t		tmp_cid_buf[OFI_CONTEXT_ID_SIZE];
-};
-
-struct util_coll_comp_item {
-	struct util_coll_hdr	hdr;
-	enum util_coll_op_type	op_type;
-	void			*data;
-	util_coll_comp_t	comp_fn;
-};
-
 struct util_coll_mc {
 	struct fid_mc		mc_fid;
 	struct fid_ep		*ep;
-	struct util_av_set 	*av_set;
+	struct util_av_set	*av_set;
 	struct slist		barrier_list;
 	struct slist		deferred_list;
-	int 			my_rank;
-	uint16_t		cid;
-	uint16_t		tag_seq;
+	struct slist		pending_xfer_list;
+	int			my_rank;
+	uint16_t		cid_id;
+	uint16_t		cid_seq;
 	ofi_atomic32_t		ref;
 };
 
+struct util_coll_state;
+
+typedef void (*util_coll_comp_fn_t)(struct util_coll_state *state);
+
+struct util_coll_cid_data {
+	uint64_t	cid_buf[OFI_CONTEXT_ID_SIZE];
+	uint64_t	tmp_cid_buf[OFI_CONTEXT_ID_SIZE];
+};
+
+struct util_coll_state {
+	struct util_coll_hdr	hdr;
+	struct dlist_entry	entry;
+	enum util_coll_op_type	type;
+	void			*context;
+	struct util_coll_mc	*mc;
+	union data {
+		struct util_coll_cid_data join;
+		struct util_coll_cid_data barrier;
+	} data;
```

Comment:
Use of a union for 2 structures that are the same structure isn't needed.

Reply:
the thought was that future collectives will require a different data structure so I went ahead and preemptively created the union, but i'll remove for now and add back as necessary.